### PR TITLE
Samrooke/ac 604 update workflow validation

### PR DESCRIPTION
**Pull request by samrooke**
### Description
Update workflow validation

Fixes # .

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] All tests passed locally.
- [ ] [Documentation comments](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/documentation-comments) included/updated.
- [ ] User guide updated.
- [ ] I have updated the changelog

### CODE REVIEW

File: src/WorkflowManager/WorkflowManager/Validators/WorkflowValidator.cs [hunk 1]
```diff
@@ -58,27 +71,26 @@ public static void Reset()
...
 
-            ValidateWorkflowSpec(workflow);
+            await ValidateWorkflowSpec(workflow);
             DetectUnreferencedTasks(tasks, firstTask);
             ValidateTask(tasks, firstTask, 0);
             ValidateTaskDestinations(workflow);
             ValidateExportDestinations(workflow);
 
-            results = (Errors.ToList(), SuccessfulPaths.ToList());
+            var results = (Errors.ToList(), SuccessfulPaths.ToList());
```
Comment (lillie-dae) [inline src/WorkflowManager/WorkflowManager/Validators/WorkflowValidator.cs:88]: is the .ToList needed here?

File: tests/UnitTests/WorkflowManager.Tests/Validators/WorkflowValidatorTests.cs [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+        public WorkflowValidatorTests()
+        {
+            _workflowService = new Mock<IWorkflowService>();
+            this.WorkflowValidator = new WorkflowValidator(_workflowService.Object);
+        }
+
+        [Fact]
+        public async void ValidateWorkflow_ValidatesAWorkflow_ReturnsErrorsAndHasCorrectValidationResultsAsync()
+        {
+            var workflow = new Workflow
```
Comment (lillie-dae) [inline tests/UnitTests/WorkflowManager.Tests/Validators/WorkflowValidatorTests.cs:25]: I think its important to have the 

            for (var i = 0; i < 15; i++)
            {
            ...
            }

on the tests to ensure reset works correctly, it ensures memory is being cleaned up correctly on workflow validator reset.

File: src/WorkflowManager/WorkflowManager/Validators/WorkflowValidator.cs (hunk 1 above)
Comment (samrooke) [reply src/WorkflowManager/WorkflowManager/Validators/WorkflowValidator.cs:88]: Yes it is, I removed this while testing something and forgot to add it back. Good spot, thank you.

File: tests/UnitTests/WorkflowManager.Tests/Validators/WorkflowValidatorTests.cs (hunk 2 above)
Comment (samrooke) [reply tests/UnitTests/WorkflowManager.Tests/Validators/WorkflowValidatorTests.cs:25]: Ah I wondered why it was doing this, - I removed it as it was doing extra work without any reason to. I'll add it back with a comment to document its purpose.

File: src/WorkflowManager/WorkflowManager/Monai.Deploy.WorkflowManager.csproj [hunk 3]
```diff
@@ -60,6 +60,7 @@
   <ItemGroup>
     <ProjectReference Include="..\..\Shared\Configuration\Monai.Deploy.WorkflowManager.Configuration.csproj" />
     <ProjectReference Include="..\..\Shared\Shared\Monai.Deploy.WorkflowManager.Shared.csproj" />
+    <ProjectReference Include="..\..\TaskManager\Plug-ins\AideClinicalReview\Monai.Deploy.WorkflowManager.TaskManager.AideClinicalReview.csproj" />
```
Comment (jackschofield23) [inline src/WorkflowManager/WorkflowManager/Monai.Deploy.WorkflowManager.csproj:63]: Don't think this should be here

File: tests/IntegrationTests/TaskManager.IntegrationTests/Monai.Deploy.WorkflowManager.TaskManager.IntegrationTests.csproj [hunk 4]
```diff
@@ -40,6 +40,7 @@
     <ProjectReference Include="..\..\..\src\TaskManager\API\Monai.Deploy.WorkflowManager.TaskManager.API.csproj" />
     <ProjectReference Include="..\..\..\src\TaskManager\Plug-ins\AideClinicalReview\Monai.Deploy.WorkflowManager.TaskManager.AideClinicalReview.csproj" />
     <ProjectReference Include="..\..\..\src\TaskManager\TaskManager\Monai.Deploy.WorkflowManager.TaskManager.csproj" />
+    <ProjectReference Include="..\..\..\src\WorkflowManager\WorkflowManager\Monai.Deploy.WorkflowManager.csproj" />
```
Comment (jackschofield23) [inline tests/IntegrationTests/TaskManager.IntegrationTests/Monai.Deploy.WorkflowManager.TaskManager.IntegrationTests.csproj:43]: should this be here also?

File: src/WorkflowManager/WorkflowManager/Monai.Deploy.WorkflowManager.csproj (hunk 3 above)
Comment (samrooke) [reply src/WorkflowManager/WorkflowManager/Monai.Deploy.WorkflowManager.csproj:63]: We need this as we're validating against required clinical review keys in the `args` param and the `Keys` constants are in this project.

File: tests/IntegrationTests/TaskManager.IntegrationTests/Monai.Deploy.WorkflowManager.TaskManager.IntegrationTests.csproj (hunk 4 above)
Comment (samrooke) [reply tests/IntegrationTests/TaskManager.IntegrationTests/Monai.Deploy.WorkflowManager.TaskManager.IntegrationTests.csproj]: Removed this one, not sure why I added it here.

File: src/WorkflowManager/WorkflowManager/Monai.Deploy.WorkflowManager.csproj (hunk 3 above)
Comment (jackschofield23) [reply src/WorkflowManager/WorkflowManager/Monai.Deploy.WorkflowManager.csproj:63]: Hmm I think its still better that we keep the projects separate and we duplicate these keys somewhere within the workflow manager.

File: src/WorkflowManager/WorkflowManager/Validators/WorkflowValidator.cs [hunk 5]
```diff
@@ -194,17 +213,27 @@ private static void ValidateTask(TaskObject[] tasks, TaskObject currentTask, int
...
                 var allOutputsUnique = currentTask.Artifacts.Output.All(x => uniqueOutputNames.Add(x.Name));
 
-                if (!allOutputsUnique)
+                if (allOutputsUnique is false)
                 {
                     Errors.Add($"Task: \"{currentTask.Id}\" has multiple output names with the same value.\n");
                 }
             }
 
+            if (currentTask.Type.Equals("argo", StringComparison.OrdinalIgnoreCase) is true)
```
Comment (rowellx68) [inline src/WorkflowManager/WorkflowManager/Validators/WorkflowValidator.cs:227]:
```suggestion
            if (currentTask.Type.Equals("argo", StringComparison.OrdinalIgnoreCase))
```

File: src/WorkflowManager/WorkflowManager/Validators/WorkflowValidator.cs [hunk 6]
```diff
@@ -194,17 +213,27 @@ private static void ValidateTask(TaskObject[] tasks, TaskObject currentTask, int
...
+            {
+                ValidateArgoTask(currentTask);
+            }
+
+            if (currentTask.Type.Equals("clinical-review", StringComparison.OrdinalIgnoreCase) is true)
```
Comment (rowellx68) [inline src/WorkflowManager/WorkflowManager/Validators/WorkflowValidator.cs:232]:
```suggestion
            if (currentTask.Type.Equals("clinical-review", StringComparison.OrdinalIgnoreCase))
```

Could probably simplify it

File: src/WorkflowManager/WorkflowManager/Validators/WorkflowValidator.cs [hunk 7]
```diff
@@ -148,13 +162,18 @@ private static void DetectUnreferencedTasks(TaskObject[] tasks, TaskObject first
             }
         }
 
-        private static void ValidateWorkflowSpec(Workflow workflow)
+        private async Task ValidateWorkflowSpec(Workflow workflow)
         {
-            if (string.IsNullOrWhiteSpace(workflow.Name))
+            if (string.IsNullOrWhiteSpace(workflow.Name) is true)
```
Comment (rowellx68) [inline src/WorkflowManager/WorkflowManager/Validators/WorkflowValidator.cs:167]: Ny reason why we added `is true` here?

Comment (rowellx68) [reply src/WorkflowManager/WorkflowManager/Validators/WorkflowValidator.cs:167]: if that's the case then ignore the other comments regarding this type of change.

Is this something we could enforce as a code style?

Comment (samrooke) [reply src/WorkflowManager/WorkflowManager/Validators/WorkflowValidator.cs:167]: As in have some form of rule that could pick up when `is true` isn't being used?

**Review by jackschofield23 — APPROVED** [review]

**Review by rowellx68 — APPROVED** [review]
lgtm

**Pull request merged**